File: Source/Hosts/PepAppHostNet/PepAppNet.cpp
/////////////////////////////////////////////////////////////////////////////
//  Copyright (C) 2019-2021 Kevin Eshbach
/////////////////////////////////////////////////////////////////////////////

#include "stdafx.h"

#include "PepAppNet.h"

#using <System.Windows.Forms.dll>

#using "UtilNet.dll"
#using "UiUtilNet.dll"
#using "UtPepNet.dll"
#using "PepForms.dll"

#pragma region "Constants"

#define CRegistryKey L"Software\\Kevin Eshbach\\PepApp\\3.00"

#define CFormLocationsName L"FormLocations"

#pragma endregion

static System::Boolean lReset(void)
{
    System::Boolean bResult = false;
    Microsoft::Win32::RegistryKey^ RegKey = Common::Registry::OpenCurrentUserRegKey(CRegistryKey, true);

    if (RegKey == nullptr)
    {
        return true;
    }

    try
    {
        for each(System::String ^ sSubKeyName in RegKey->GetSubKeyNames())
        {
            RegKey->DeleteSubKeyTree(sSubKeyName);
        }

        bResult = true;
    }
    finally
    {
        RegKey->Close();
    }

    return bResult;
}

Pep::Application::Startup::Startup()
{
}

System::UInt32 Pep::Application::Startup::Execute(
  System::Boolean bUseParallelPort,
  System::Boolean bReset)
{
	System::String^ sFormLocationsRegistryKey = System::String::Format(L"{0}\\{1}", CRegistryKey, CFormLocationsName);
    Pep::Programmer::Config::EDeviceType DeviceType = Pep::Programmer::Config::EDeviceType::USB;
    Pep::Forms::MainForm^ AppForm;

    if (bUseParallelPort)
    {
        DeviceType = Pep::Programmer::Config::EDeviceType::ParallelPort;
    }

    if (bReset)
    {
        return lReset() ? 0 : 1;
    }

	if (!Common::IO::TempFileManager::Initialize())
	{
	}

    try
    {
        System::Windows::Forms::Application::EnableVisualStyles();
        System::Windows::Forms::Application::SetCompatibleTextRenderingDefault(false);

        AppForm = gcnew Pep::Forms::MainForm(CRegistryKey, sFormLocationsRegistryKey);

		gcnew Common::Forms::FormLocation(AppForm, sFormLocationsRegistryKey);

		Pep::Programmer::Config::Initialize(DeviceType, AppForm);

        Common::Forms::Application::Run(AppForm);

		Pep::Programmer::Config::Uninitialize();
	}
    catch (System::Exception^ exception)
    {
        System::String^ sMsg;

        sMsg = System::String::Format(L"Unhandled exception.  ({0})\n\nThe application will now automatically close to prevent data loss.", 
                                      exception->Message);

        System::Windows::Forms::MessageBox::Show(sMsg,
            System::Windows::Forms::Application::ProductName,
            System::Windows::Forms::MessageBoxButtons::OK,
            System::Windows::Forms::MessageBoxIcon::Error);
    }

	if (!Common::IO::TempFileManager::Uninitialize())
	{
	}

    return 0;
}

/////////////////////////////////////////////////////////////////////////////
//  Copyright (C) 2019-2021 Kevin Eshbach
/////////////////////////////////////////////////////////////////////////////

File: Source/AppsNet/PepAppNet/PepAppNet.cs
/***************************************************************************/
/*  Copyright (C) 2006-2014 Kevin Eshbach                                  */
/***************************************************************************/

using System;

namespace Pep
{
    namespace Application
    {
        static class Startup
        {
            /// <summary>
            /// The main entry point for the application.
            /// </summary>
            [STAThread]
            static void Main()
            {
                Common.Process.Running AppRunning;
                Pep.Forms.MainForm AppForm;

                AppRunning = new Common.Process.Running(System.Windows.Forms.Application.ProductName);

                try
                {
                    System.Windows.Forms.Application.EnableVisualStyles();
                    System.Windows.Forms.Application.SetCompatibleTextRenderingDefault(false);

                    if (AppRunning.AlreadyRunning == false)
                    {
                        Pep.Controls.Config.Initialize();

                        AppForm = new Pep.Forms.MainForm();

                        System.Windows.Forms.Application.AddMessageFilter(AppForm);

                        Common.Forms.Application.Run(AppForm);

                        System.Windows.Forms.Application.RemoveMessageFilter(AppForm);

                        Pep.Controls.Config.Uninitialize();
                    }
                    else
                    {
                        System.Windows.Forms.MessageBox.Show("The application is already running.",
                            System.Windows.Forms.Application.ProductName,
                            System.Windows.Forms.MessageBoxButtons.OK,
                            System.Windows.Forms.MessageBoxIcon.Error);
                    }
                }
                catch (Exception e)
                {
                    System.String sMsg;

                    sMsg = "Unhandled exception caught.  (" + e.Message +
                        ")\n\nThe application will now automatically close to prevent data loss.";

                    System.Windows.Forms.MessageBox.Show(sMsg,
                        System.Windows.Forms.Application.ProductName,
                        System.Windows.Forms.MessageBoxButtons.OK,
                        System.Windows.Forms.MessageBoxIcon.Error);
                }

                AppRunning.Close();
            }
        }
    }
}

/***************************************************************************/
/*  Copyright (C) 2006-2014 Kevin Eshbach                                  */
/***************************************************************************/
